### ISSUE TITLE
shady dom is unusable (doesn't render children in many instances), which hurts Web Component uptake

### ISSUE THREAD

**Issue by knewter** [elm-lang/virtual-dom#65]
## tl;dr

Presently (as of 0.18), to use web components with Elm requires the full shadow DOM polyfill (or else child nodes aren't rendered in many cases).  The Shadow DOM polyfill yields poor performance everywhere but Chrome.

This will hurt adoption of Web Components in Elm, but Web Components make it easier to move stateful UI bits out and seem to solve the elm-mdl ripple problem nicely.

I'd be interested in trying to help fix this issue, and just wanted to make sure that I could get some guidance or suggestions on whether the work would be in vain (i.e. if this problem is already on the radar / being worked on). Thoughts?

## More detailed explanation

- When using the Shady DOM polyfill (the default for the webcomponents polyfill), children aren't rendered in many instances.  If you use a Polymer App layout, it doesn't render the content of the app at all!
- When switching to Shadow DOM (with `window.Polymer = { dom: 'shadow' }`) everything renders nicely but performance is poor in Firefox and Safari Mobile, at least.

## Initial thought re: where to look

- timg on slack mentioned that shadydom requires you to notify it when new dom elements have been added.  This likely requires a hook of some sort.  I'll try to update this issue with documentation on how that works as I find it.
- Additionally, I'll try this out with Polymer 2's beta just in case it's better there - but I'm doubtful.

## Reproduction

Here's a git repo reproducing the issue with polymer web components: https://github.com/knewter/elm_web_components_playground/commit/72cdde0e38030bf694fe1c85c229b2f049d72cce

it's not an SSCCE yet, but the view function itself is pretty easy to comprehend: https://github.com/knewter/elm_web_components_playground/blob/72cdde0e38030bf694fe1c85c229b2f049d72cce/src/App.elm#L36

If you open up `src/index.html` and uncomment the shadow dom configuration, it works fine - but the shadow dom polyfill has poor performance everywhere but Chrome.  If you open up the app in firefox and click the login button, you'll note that the ripple doesn't show up because firefox performs slowly with the full shadow dom polyfill.  We have to use that polyfill at present, because otherwise we don't even render the app in this example!

## Resources

- [Shady DOM explained](https://www.polymer-project.org/1.0/blog/shadydom)
  - From the above: `Many projects simply cannot afford the Shadow DOM Polyfill, for the reasons given above. In particular, the performance costs on platforms like mobile Safari are nearly intolerable.`
- https://github.com/webcomponents/shadydom
- [Here's a Pull Request (and corresponding discussion) that adds support to Matt-Esch/virtual-dom](https://github.com/Matt-Esch/virtual-dom/pull/296)

**Comment by knewter** [comment]
After [some discussion on elm-dev](https://groups.google.com/forum/#!topic/elm-dev/TvnVOpz3l0Q), it seems that this is unnecessary and can be handled by switching out the implementation of shadow dom based on the browser used - firefox and safari mobile work fine with shady dom, only chrome dies when using shady dom and I can just do a browser check.  Thanks!

**Comment by knewter** [comment]
Sigh maybe not...click events aren't being bubbled up under shadydom on firefox.  Digging in further...

**Comment by rtfeldman** [comment]
Hm...I hate to say this, but does an Elm solution make sense here?

It seems like the problem is that the official Web Components polyfill is unreliable.

There might be ways for Elm to improve that situation...but what if there aren't? There's a risk of spending a bunch of time trying to improve compatibility with this polyfill, only to find that it's still too unreliable.

Maybe the answer is that despite how hard Google has been pushing the opposite narrative, Polymer just isn't ready for normal use until browsers natively support Shadow DOM. 😕

**Comment by knewter** [comment]
@rtfeldman I really think it's all in how the elm virtual-dom interacts with shadydom.  ShadyDOM works fine in general - it just doesn't play nicely with elm's virtual-dom as far as I can tell.

I'm really trying here :)  We don't want elm-mdl style ripples, and web components are a likely solution - cool!  They work great and in fact seem to open us up to being able to **use a *ton* of awesome things**, so that's a win.

But then there's this issue.  I believe it's likely tractable and would be willing to put in the time researching / attempting further, if it had the chance of seeing the light of day.

Right now, if you want to use web components with elm and support iOS devices, you have bad performance.  My guess is that would be true for a couple of years.

One thing - this presentation suggests that the shadydom problem I'm seeing can be solved with `Html.Keyed`, which I hadn't considered (but should have) - http://slides.com/kevinlebrun/elm-polymer#/5/3

If that works, I'm back to suggesting Html.Keyed at the primary screen-replace boundary, and writing a quick browser-test to determine how to load polymer.  Testing now.

**Comment by knewter** [comment]
My experiments were problematic at best.  I never got to anything usable using Keyed :-\  My guess is changes between 0.17 virtual-dom and 0.18 virtual-dom (if that presentation worked as it suggested - the code I pulled mostly kind of worked right on 0.17).

**Comment by rtfeldman** [comment]
Hm...I doubt it's a recent thing, given the original vdom library elm's
originally used is seeing similar-sounding symptoms:
https://github.com/Matt-Esch/virtual-dom/pull/296

**Comment by evancz** [comment]
@lukewestby and I looked into Shady DOM today.

As far as we can tell, it is fine to use web components as long as they have no child nodes that are owned by Elm. In other words, they are leafs in the virtual DOM structure. There are a few ways to go from here:

  1. Change `virtual-dom` to know about this when it traverses things. This should be possible assuming (1) nodes are marked as "shadow DOM nodes" when web components create them and (2) web components do not modify *our* nodes.

  2. Create a website that lists the web components that already work with Elm. If there are any that do everything through attributes, those will be fine with shady DOM as is. I think it'd be valuable to have a list of these, so people can assess if *enough* are available for their scenario.

  3. Wrap existing web components in an API that makes them look like leafs. Maybe that means having *it* manage its children separately from the root of your application. In other words, each component can do its own diff and patch on the appropriate things.

Based on these routes, I think the next steps are:

  1. Create a repo where the README lists web components that already work. Share the link here.
  2. Find out if "shady" nodes are marked in some way. If so, how? Explain here.

@knewter, can you do those two things?

**Comment by evancz** [comment]
Ultimately, I think we should have a website that explains how to use all this stuff with examples. That means showing Elm code instead of raw HTML, and presenting all Elm specific information so people do not need to translate and know about tricky details.

So I think making the list is work that will be needed eventually no matter what.
